File: exercises/python/shared_observability/kafka_header_propagator.py
"""Propagate domain identifiers through Kafka message headers."""

from __future__ import annotations

from typing import Sequence

from shared_observability.domain_identifier import DomainIdentifier

# Prefix used to namespace domain identifiers in Kafka headers.
_HEADER_PREFIX = "domain."


def inject_domain_identifiers(
    identifiers: Sequence[DomainIdentifier],
) -> list[tuple[str, bytes]]:
    """Convert domain identifiers to Kafka headers.

    Returns a list of (header_name, header_value_bytes) tuples ready to
    pass to a Kafka producer's ``headers`` parameter.
    """
    return [
        (f"{_HEADER_PREFIX}{ident.key()}", ident.value().encode("utf-8"))
        for ident in identifiers
    ]


def extract_domain_identifiers(
    headers: list[tuple[str, bytes]] | None,
) -> dict[str, str]:
    """Extract domain identifiers from Kafka headers.

    Returns a dict mapping identifier keys to their string values.
    Only headers prefixed with ``domain.`` are considered.
    """
    if not headers:
        return {}

    result: dict[str, str] = {}
    for name, value in headers:
        if name.startswith(_HEADER_PREFIX):
            key = name[len(_HEADER_PREFIX) :]
            result[key] = value.decode("utf-8")
    return result

File: exercises/dotnet/SharedObservability/KafkaHeaderPropagator.cs
using System.Text;

namespace SharedObservability;

/// <summary>
/// Propagates <see cref="IDomainIdentifier"/> instances to and from Kafka
/// message headers, enabling domain context to flow through event-driven
/// communication without coupling to OTel wire format.
/// </summary>
public static class KafkaHeaderPropagator
{
    private const string HeaderPrefix = "domain.";

    /// <summary>
    /// Writes domain identifiers into Kafka headers. Each identifier is stored
    /// as a UTF-8 header with key "domain.{identifier.Key}".
    /// </summary>
    /// <param name="headers">The Kafka headers collection (list of key-value byte pairs).</param>
    /// <param name="identifiers">Domain identifiers to propagate.</param>
    public static void Inject(IList<KeyValuePair<string, byte[]>> headers, params IDomainIdentifier[] identifiers)
    {
        foreach (var id in identifiers)
        {
            var headerKey = HeaderPrefix + id.Key;
            headers.Add(new KeyValuePair<string, byte[]>(headerKey, Encoding.UTF8.GetBytes(id.Value)));
        }
    }

    /// <summary>
    /// Reads domain identifier values from Kafka headers. Returns a dictionary
    /// of identifier keys (without the "domain." prefix) to their values.
    /// </summary>
    /// <param name="headers">The Kafka headers collection.</param>
    /// <returns>Dictionary of domain identifier key-value pairs found in headers.</returns>
    public static IReadOnlyDictionary<string, string> Extract(IEnumerable<KeyValuePair<string, byte[]>> headers)
    {
        var result = new Dictionary<string, string>();

        foreach (var header in headers)
        {
            if (header.Key.StartsWith(HeaderPrefix, StringComparison.Ordinal))
            {
                var domainKey = header.Key[HeaderPrefix.Length..];
                result[domainKey] = Encoding.UTF8.GetString(header.Value);
            }
        }

        return result;
    }
}
